File: packages/core/src/tokens.ts
// Token counting from API responses + estimation

import type { TokenUsage } from "./types.js"

const ANTHROPIC_USAGE =
  /"usage"\s*:\s*\{[^}]*"input_tokens"\s*:\s*(\d+)[^}]*"output_tokens"\s*:\s*(\d+)/
const OPENAI_USAGE =
  /"usage"\s*:\s*\{[^}]*"prompt_tokens"\s*:\s*(\d+)[^}]*"completion_tokens"\s*:\s*(\d+)/

/**
 * Parse token usage from a JSON response body.
 * Supports Anthropic and OpenAI formats.
 * Returns zero usage when parsing fails — never throws.
 */
export function parseTokenUsage(responseBody: string): TokenUsage {
  const anthropicMatch = ANTHROPIC_USAGE.exec(responseBody)
  if (anthropicMatch?.[1] && anthropicMatch[2]) {
    const input = parseInt(anthropicMatch[1], 10)
    const output = parseInt(anthropicMatch[2], 10)
    return { inputTokens: input, outputTokens: output, totalTokens: input + output }
  }

  const openaiMatch = OPENAI_USAGE.exec(responseBody)
  if (openaiMatch?.[1] && openaiMatch[2]) {
    const input = parseInt(openaiMatch[1], 10)
    const output = parseInt(openaiMatch[2], 10)
    return { inputTokens: input, outputTokens: output, totalTokens: input + output }
  }

  return { inputTokens: 0, outputTokens: 0, totalTokens: 0 }
}

/**
 * Estimate token count from text (~4 characters per token).
 */
export function estimateTokens(text: string): number {
  return Math.ceil(text.length / 4)
}

const PRICE_PER_MILLION: Record<string, number> = {
  "api.anthropic.com": 3.0,
  "api.openai.com": 2.5,
  "api.githubcopilot.com": 0.0,
  "api2.cursor.sh": 0.0,
  default: 3.0,
}

export function estimateCost(tokensSaved: number, upstream: string): number {
  const price = PRICE_PER_MILLION[upstream] ?? PRICE_PER_MILLION["default"] ?? 3.0
  return (tokensSaved / 1_000_000) * price
}

File: go/internal/trimmer/tokens.go
package trimmer

import (
	"encoding/json"
	"strings"
)

// UsageAnthropic matches Anthropic API response usage field.
type UsageAnthropic struct {
	InputTokens  int `json:"input_tokens"`
	OutputTokens int `json:"output_tokens"`
}

// UsageOpenAI matches OpenAI API response usage field.
type UsageOpenAI struct {
	PromptTokens     int `json:"prompt_tokens"`
	CompletionTokens int `json:"completion_tokens"`
	TotalTokens      int `json:"total_tokens"`
}

// TokenUsage holds normalized token counts from any API response.
type TokenUsage struct {
	InputTokens  int
	OutputTokens int
	TotalTokens  int
}

// ParseTokenUsage extracts token counts from an API response body.
func ParseTokenUsage(responseBody []byte) TokenUsage {
	var anthropic struct {
		Usage UsageAnthropic `json:"usage"`
	}
	if err := json.Unmarshal(responseBody, &anthropic); err == nil {
		if anthropic.Usage.InputTokens > 0 || anthropic.Usage.OutputTokens > 0 {
			return TokenUsage{
				InputTokens:  anthropic.Usage.InputTokens,
				OutputTokens: anthropic.Usage.OutputTokens,
				TotalTokens:  anthropic.Usage.InputTokens + anthropic.Usage.OutputTokens,
			}
		}
	}

	var openai struct {
		Usage UsageOpenAI `json:"usage"`
	}
	if err := json.Unmarshal(responseBody, &openai); err == nil {
		if openai.Usage.TotalTokens > 0 {
			return TokenUsage{
				InputTokens:  openai.Usage.PromptTokens,
				OutputTokens: openai.Usage.CompletionTokens,
				TotalTokens:  openai.Usage.TotalTokens,
			}
		}
	}

	return TokenUsage{}
}

// ParseStreamTokenUsage extracts token usage from SSE event lines.
func ParseStreamTokenUsage(events []string) TokenUsage {
	for i := len(events) - 1; i >= 0; i-- {
		line := events[i]
		if !strings.HasPrefix(line, "data: ") {
			continue
		}
		data := strings.TrimPrefix(line, "data: ")
		if data == "[DONE]" {
			continue
		}

		usage := ParseTokenUsage([]byte(data))
		if usage.TotalTokens > 0 {
			return usage
		}
	}
	return TokenUsage{}
}

// EstimateTokens gives a rough estimate when no usage data is available.
func EstimateTokens(text []byte) int {
	if len(text) == 0 {
		return 0
	}
	return len(text) / 4
}
